File: LeetCode-Cpp/CourseScheduleII.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
         // Adjacency list ve in-degree dizisi
        vector<vector<int>> adj(numCourses);
        vector<int> indeg(numCourses, 0);

        // Kenarlar: b -> a (a için önce b alınmalı)
        for (const auto& p : prerequisites) {
            int a = p[0], b = p[1];
            adj[b].push_back(a);
            indeg[a]++;
        }

        // in-degree'i 0 olan dersleri kuyruğa at
        queue<int> q;
        for (int i = 0; i < numCourses; ++i)
            if (indeg[i] == 0) q.push(i);

        vector<int> order;
        order.reserve(numCourses);

        // BFS
        while (!q.empty()) {
            int u = q.front(); q.pop();
            order.push_back(u);
            for (int v : adj[u]) {
                if (--indeg[v] == 0) q.push(v);
            }
        }

        // Tüm dersler eklendiyse geçerli sıra; yoksa döngü var demektir
        if ((int)order.size() == numCourses) return order;
        return {};
    }
};

File: LeetCode-Python/CourseScheduleII.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
         # Graf: b -> a (a'yı almak için önce b)
        adj = defaultdict(list)
        indeg = [0] * numCourses

        for a, b in prerequisites:
            adj[b].append(a)
            indeg[a] += 1

        # Önkoşulu olmayan dersler kuyruğa
        q = deque([i for i in range(numCourses) if indeg[i] == 0])
        order = []

        while q:
            u = q.popleft()
            order.append(u)
            for v in adj[u]:
                indeg[v] -= 1
                if indeg[v] == 0:
                    q.append(v)

        # Tüm dersler eklendiyse geçerli sıra, yoksa döngü var -> boş liste
        return order if len(order) == numCourses else []
